torchvision/prototype/models/vision_transformer.py

```diff
@@ -0,0 +1,383 @@
+        else:
+            heads_layers["pre_logits"] = nn.Linear(hidden_dim, representation_size)
+            heads_layers["act"] = nn.Tanh()
+            heads_layers["head"] = nn.Linear(representation_size, num_classes)
+
+        self.heads = nn.Sequential(heads_layers)
+        self._init_weights()
+
+    def _init_weights(self):
+        fan_in = self.conv_proj.in_channels * self.conv_proj.kernel_size[0] * self.conv_proj.kernel_size[1]
+        nn.init.trunc_normal_(self.conv_proj.weight, std=math.sqrt(1 / fan_in))
+        nn.init.zeros_(self.conv_proj.bias)
+
+        if hasattr(self.heads, "pre_logits"):
+            fan_in = self.heads.pre_logits.in_features
+            nn.init.trunc_normal_(self.heads.pre_logits.weight, std=math.sqrt(1 / fan_in))
+            nn.init.zeros_(self.heads.pre_logits.bias)
+
+        nn.init.zeros_(self.heads.head.weight)
+        nn.init.zeros_(self.heads.head.bias)
+
+    def forward(self, x: torch.Tensor):
+        n, c, h, w = x.shape
+        p = self.patch_size
+        torch._assert(h == self.image_size, "Wrong image height!")
+        torch._assert(w == self.image_size, "Wrong image width!")
+        n_h = h // p
+        n_w = w // p
+
+        # (n, c, h, w) -> (n, hidden_dim, n_h, n_w)
+        x = self.conv_proj(x)
+        # (n, hidden_dim, n_h, n_w) -> (n, hidden_dim, (n_h * n_w))
+        x = x.reshape(n, self.hidden_dim, n_h * n_w)
+
+        # (n, hidden_dim, (n_h * n_w)) -> (n, (n_h * n_w), hidden_dim)
+        # The self attention layer expects inputs in the format (N, S, E)
+        # where S is the source sequence length, N is the batch size, E is the
+        # embedding dimension
+        x = x.permute(0, 2, 1)
+
+        # Expand the class token to the full batch.
+        batch_class_token = self.class_token.expand(n, -1, -1)
+        x = torch.cat([batch_class_token, x], dim=1)
+
+        x = self.encoder(x)
+
+        # Classifier "token" as used by standard language architectures
+        x = x[:, 0]
+
+        x = self.heads(x)
+
+        return x
+
+
+class VisionTransformer_B_16Weights(Weights):
+    # If a default model is added here the corresponding changes need to be done in vit_b_16
+    pass
+
+
+class VisionTransformer_B_32Weights(Weights):
+    # If a default model is added here the corresponding changes need to be done in vit_b_32
+    pass
+
+
+class VisionTransformer_L_16Weights(Weights):
+    # If a default model is added here the corresponding changes need to be done in vit_l_16
+    pass
+
+
+class VisionTransformer_L_32Weights(Weights):
+    # If a default model is added here the corresponding changes need to be done in vit_l_32
+    pass
+
+
+def _vision_transformer(weights: Optional[Weights], progress: bool, **kwargs: Any) -> VisionTransformer:
+    image_size = kwargs.get("image_size", 224)
+    if "image_size" in kwargs:
+        kwargs.pop("image_size")
+
+    model = VisionTransformer(image_size=image_size, **kwargs)
```

Comment:
Optional nit: Given the parameters on the constructor are value arguments and not named ones, you could consider passing their values explicitly instead via the `**kwargs`. This is what we usually do for other models, for example:
https://github.com/pytorch/vision/blob/9841a9076e711f73482cb84e106fe5ec40707023/torchvision/prototype/models/mobilenetv3.py#L22-L32

There are a few discussions about updating our signatures on the future (see https://github.com/pytorch/vision/pull/4937#discussion_r749549044) so this could future-proof the implementation.

Reply:
What do you mean by `passing their values explicitly instead via the **kwargs`? In the above example, `num_classes` is passed via `**kwargs`, right?

Reply:
Sorry, I should have been clearer. 

In the MobileNetV3 example above, I'm referring to the model configuration value parameters passed by the model builder. That is the `inverted_residual_setting` and `last_channel`. 

In your case, you could explicitly pass the parameters `patch_size`, `num_layers`, `num_heads`, `hidden_dim` and `mlp_dim`. All of them (including `image_size`) are not named and are always passed. The `kwargs` in our setup stores optional named parameters for the constructor which have default values.

I should note, this is really a nitpick and definitely not a blocking one.

Reply:
Got it! Let me make changes accordingly :)